### PR TITLE
gh-103092: Prep curses module for multi-phase init

### PR DESCRIPTION
**Pull request by koubaa**
https://bugs.python.org/issue1635741

* Issue: gh-103092

### CODE REVIEW

File: Modules/_cursesmodule.c [hunk 1]
```diff
@@ -3916,13 +3927,18 @@ _curses_qiflush_impl(PyObject *module, int flag)
...
+update_lines_cols(PyObject *module)
 {
     PyObject *o;
     PyObject *m = PyImport_ImportModuleNoBlock("curses");
     _Py_IDENTIFIER(LINES);
     _Py_IDENTIFIER(COLS);
 
+    PyObject *d = PyModule_GetDict(module);
+    if (d == NULL) {
+        return NULL;
```
Comment (shihai1991) [inline Modules/_cursesmodule.c]: > return makes integer from pointer without a cast [-Wint-conversion]

wrong return type in here?

File: Modules/_cursesmodule.c [hunk 2]
```diff
@@ -4681,25 +4699,25 @@ static PyMethodDef PyCurses_methods[] = {
...
-    -1,
-    PyCurses_methods,
-    NULL,
-    NULL,
-    NULL,
-    NULL
+    .m_name = "_curses",
+    .m_doc = NULL,
+    .m_size = -1,
+    .m_nethods = PyCurses_methods,
```
Comment (shihai1991) [inline Modules/_cursesmodule.c]: Another Error report in here too.
typo error.

File: Modules/_cursesmodule.c [hunk 3]
```diff
@@ -4681,25 +4699,25 @@ static PyMethodDef PyCurses_methods[] = {
...
+    .m_slots = NULL,
```
Comment (shihai1991) [inline Modules/_cursesmodule.c:4730]: Remove them?

**Comment by shihai1991** [comment]
cc @vstinner @corona10

File: Modules/_cursesmodule.c (hunk 1 above)
Comment (koubaa) [reply Modules/_cursesmodule.c]: @shihai1991 Sorry, this is not ready to review.  I don't have an environment to test curses (I use either windows or WSL for linux and neither can build/test it) so I am using the CI to test it.

File: Modules/_cursesmodule.c (hunk 3 above)
Comment (koubaa) [reply Modules/_cursesmodule.c:4730]: @shihai1991 I will re-add in the next PR (to change to multi-phase init) so I thought I might as well add them now and keep as NULL.

File: Modules/_cursesmodule.c (hunk 1 above)
Comment (shihai1991) [reply Modules/_cursesmodule.c]: This all right. I suggest you can create a PR to your fork repo. The fork repo's PR can trigger the CI gate too.

Comment (koubaa) [reply Modules/_cursesmodule.c]: @shihai1991 I'll try that for the next PR which actually converts to multi-phase init.  It seems like this one is done.

**Comment by koubaa** [comment]
@vstinner @corona10 @shihai1991 please review

File: Modules/_cursesmodule.c [hunk 4]
```diff
@@ -3916,13 +3927,18 @@ _curses_qiflush_impl(PyObject *module, int flag)
  * and _curses.COLS */
 #if defined(HAVE_CURSES_RESIZETERM) || defined(HAVE_CURSES_RESIZE_TERM)
 static int
-update_lines_cols(void)
+update_lines_cols(PyObject *module)
 {
     PyObject *o;
     PyObject *m = PyImport_ImportModuleNoBlock("curses");
```
Comment (vstinner) [inline Modules/_cursesmodule.c]: Why is the module still imported if you get it as an argument?

Comment (koubaa) [reply Modules/_cursesmodule.c]: @vstinner I didn't even notice this.  I added it as an argument for the PyModule_GetDict call

**Comment by koubaa** [comment]
@vstinner would you please review?

File: Modules/_cursesmodule.c [hunk 5]
```diff
@@ -3271,9 +3274,14 @@ _curses_initscr_impl(PyObject *module)
 #define SetDictInt(string,ch)                                           \
     do {                                                                \
         PyObject *o = PyLong_FromLong((long) (ch));                     \
-        if (o && PyDict_SetItemString(ModDict, string, o) == 0)     {   \
+        if (o == NULL) {                                                \
+            goto error;                                                 \
+        }                                                               \
+        if (PyDict_SetItemString(d, string, o) < 0) {                   \
             Py_DECREF(o);                                               \
+            goto error;                                                 \
```
Comment (vstinner) [inline Modules/_cursesmodule.c:3282]: "return NULL;" is enough: you should remove the error label. An error label is used when you need to clean up things, it's not the case here. Same remark at the bottom of the file.

File: Modules/_cursesmodule.c [hunk 6]
```diff
@@ -4738,22 +4744,24 @@ curses_destructor(PyObject *op)
...
 {
-    PyObject *m, *d, *v, *c_api_object;
-
     /* Initialize object type */
-    if (PyType_Ready(&PyCursesWindow_Type) < 0)
+    if (PyType_Ready(&PyCursesWindow_Type) < 0) {
         return NULL;
+    }
+
+    static void *PyCurses_API[PyCurses_API_pointers];
```
Comment (vstinner) [inline Modules/_cursesmodule.c:4752]: There are warnings on the PR, like ‘PyCurses_API’ defined but not used [-Wunused-variable].

**Comment by iritkatriel** [comment]
https://bugs.python.org/issue1635741 is closed. What is the status of this PR?

**Comment by erlend-aasland** [comment]
Closing as per https://github.com/python/cpython/issues/101714#issuecomment-1611057656. Thanks for the effort, Mohamed!

**Pull request closed without merging**